tools/plugins/com.liferay.ide.upgrade.core/src/com/liferay/blade/eclipse/provider/SSEXMLFile.java

```diff
@@ -90,4 +94,48 @@
 		return results;
 	}
 
+	@Override
+	public SearchResult getDocumentTypeDeclaration(String version, String filter) {
+		SearchResult result = null;
+
+		IFile xmlFile = getIFile(file);
+		IDOMModel domModel = null;
+
+		try (FileReader fileReader = new FileReader(file);
+			BufferedReader bufferedReader = new BufferedReader(fileReader)) {
+
+			IModelManager modelManager = StructuredModelManager.getModelManager();
+
+			domModel = (IDOMModel)modelManager.getModelForRead(xmlFile);
+
+			IDOMDocument document = domModel.getDocument();
+
+			String documentType = Objects.toString(document.getDoctype(), "");
+
+			if (documentType.contains(version) && documentType.contains(filter)) {
+				String firstLine = bufferedReader.readLine();
```

Comment:
You don't need to read this file into a buffer to fine the first/second lines.  Once you have a IDOMDocument, you can use this object model to retrieve nodes on the DOM (such as doctype) and this object can be queired to find its offset location in the file.  Ask @jtydhr88  for examples of how to do this.